src/dir/glob.cr

```diff
@@ -1,266 +1,369 @@
-        nested_path = nested_path[1..-1]
-      end
+    private def compile(pattern, patterns = [] of Array(PatternType))
+      reader = Char::Reader.new(pattern)
 
-      regex = glob2regex(ptrn)
+      lbrace = nil
+      rbrace = nil
+      alt_start = nil
 
-      scandir(dir, nested_path, regex, 0, recursion_depth, wants_dir) do |path|
-        if starts_with_path_separator
-          yield File::SEPARATOR + path
-        else
-          yield path
+      alternatives = [] of String
+
+      nest = 0
+      while char = reader.current_char
+        case char
+        when Char::ZERO
+          break
+        when '{'
+          lbrace = reader.pos if nest == 0
+          nest += 1
+        when '}'
+          nest -= 1
+
+          if nest == 0
+            rbrace = reader.pos
+            start = (alt_start || lbrace).not_nil! + 1
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            break
+          end
+        when ','
+          if nest == 1
+            start = (alt_start || lbrace).not_nil! + 1
+            alternatives << pattern.byte_slice(start, reader.pos - start)
+            alt_start = reader.pos
+          end
         end
+
+        reader.next_char
       end
-    end
-  end
 
-  private def self.glob2regex(pattern)
-    if pattern.size == 0 || pattern == File::SEPARATOR
-      raise ArgumentError.new "Empty glob pattern"
-    end
+      if lbrace && rbrace
+        front = pattern.byte_slice(0, lbrace)
+        back = pattern.byte_slice(rbrace + 1)
 
-    # characters which are escapable by a backslash in a glob pattern;
-    # Windows paths must have double backslashes:
-    escapable = {'?', '{', '}', '*', ',', '\\'}
-    # characters which must be escaped in a PCRE regex:
-    escaped = {'$', '(', ')', '+', '.', '[', '^', '|', '/'}
+        alternatives.each do |alt|
+          brace_pattern = [front, alt, back].join
 
-    last_is_file_separator = false
+          compile brace_pattern, patterns
+        end
+      else
+        patterns << single_compile pattern
+      end
 
-    regex_pattern = String.build do |str|
-      str << "\\A"
-      idx = 0
-      nest = 0
+      patterns
+    end
 
-      idx = 1 if pattern[0] == File::SEPARATOR
-      size = pattern.size
+    private def single_compile(glob)
+      list = [] of PatternType
+      return list if glob.empty?
 
-      while idx < size
-        char = pattern[idx]
+      parts = glob.split('/', remove_empty: true)
 
-        if last_is_file_separator && char == File::SEPARATOR
-          idx += 1
-          next
+      if glob.ends_with?('/')
+        list << DirectoriesOnly.new
+      else
+        file = parts.pop
+        if /^[a-zA-Z0-9._]+$/.match(file)
+          list << ConstantEntry.new file
+        elsif !file.empty?
+          list << EntryMatch.new file
         end
+      end
 
-        last_is_file_separator = char == File::SEPARATOR
+      while !parts.empty?
+        dir = parts.pop
```

Comment:
Do you really need `pop` here? I think not modifying the array will be faster. If not, you can also do `while dir = parts.pop?`

Reply:
I don't think pop is necessary.

Reply:
In fact, `StartRecursiveDirectories` was not necessary at all. It was taken from the rubinius implementation but it seems we don't need to distinguish because of a different iteration algorithm.